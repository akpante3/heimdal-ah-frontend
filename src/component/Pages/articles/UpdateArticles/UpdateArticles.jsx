import { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { toastr as feedback } from 'react-redux-toastr';
import setToken from '../../../../config/setToken';
import './UpdateArticles.scss';
import updateArticles from '../../../../actions/ArticleActions/updateArticlesActions';
import { getArticleById } from '../../../../actions/ArticleActions/getArticlesByIdActions';

require('medium-editor/dist/css/medium-editor.css');
require('medium-editor/dist/css/themes/default.css');

/**
 * @description - Helps a user updates his articles
 * @param {props} updateArticlesStatus - the status returned from dispatching updateArticles action
 * @param {props} title - the article title returned from dispatching getArticlesById action
 * @param {props} body - the article body returned from dispatching getArticlesById action
 * @param {props} updateArticles - the action been dispatched
 * @param {props} getArticlesById - the action been dispatched
 * @returns {component} Component
 */
export class UpdateArticles extends Component {
  constructor(props) {
    super(props);
    this.state = {
      updatedTitle: '',
      updatedText: '',
    };
    this.onHandleText = this.onHandleText.bind(this);
    this.onHandleTitle = this.onHandleTitle.bind(this);
    this.onHandleSubmit = this.onHandleSubmit.bind(this);
  }

  /**
 * @description - Handles the initial component rendering by dispatching getArticleById action
 * @returns {component} the initial rendered component
 */
  componentDidMount() {
    const identifier = new URLSearchParams(window.location.search).get('id');
    const token = localStorage.getItem('access-token');
    setToken(token);
    this.props.actions(getArticleById(identifier));
  }

  /**
 * @description - Handles the form submit by calling updateArticles action
 * @param {event} e - the event passed to the method
 * @returns {component} null
 */
  onHandleSubmit(e) {
    e.preventDefault();
    const identifier = new URLSearchParams(window.location.search).get('id');
    const token = localStorage.getItem('access-token');
    setToken(token);
    const { updatedTitle, updatedText } = this.state;
    if ((!updatedTitle && !updatedText)) {
      feedback.warning('Article unchanged', 'You did not update the article');
    } else if (updatedText && !updatedTitle) {
      const description = this.props.title.substring(0, 100);
      this.props.actions(updateArticles(this.props.title, updatedText, description, identifier));
    } else if (updatedText && updatedTitle) {
      const description = updatedTitle.substring(0, 100);
      this.props.actions(updateArticles(updatedTitle, updatedText, description, identifier));
    } else if (updatedTitle && !updatedText) {
      const description = updatedTitle.substring(0, 100);
      this.props.actions(updateArticles(updatedTitle, this.props.body, description, identifier));
    }
  }

  /**
 * @description - Handles the Editor body input change
 * @param {string} text - the text passed to the method
 * @returns {component} updated body state
 */
  onHandleText(text) {
    this.setState({ updatedText: text });
  }

  /**
 * @description - Handles the Editor title input change
 * @param {string} text - the text passed to the method
 * @returns {component} updated title state
 */
  onHandleTitle(text) {
    this.setState({ updatedTitle: text });
  }

  /**
 * @description - Handles the rendering of the component
 * @returns {component} the rendered component
 */
  render() {
    const { updatedText, updatedTitle } = this.state;
    const { updateArticlesStatus, title, body } = this.props;
    return this.props.children({
      updatedText,
      updatedTitle,
      title,
      body,
      updateArticlesStatus,
      onHandleSubmit: this.onHandleSubmit,
      onHandleText: this.onHandleText,
      onHandleTitle: this.onHandleTitle,
    });
  }
}

UpdateArticles.defaultProps = {
  title: 'Article Title',
  body: 'Article Body',
};

UpdateArticles.propTypes = {
  title: PropTypes.string,
  body: PropTypes.string,
  updateArticlesStatus: PropTypes.string.isRequired,
};

const mapStateToprops = state => ({
  title: state.getArticlesById.payload.title,
  body: state.getArticlesById.payload.body,
  updateArticlesStatus: state.updateArticles.status,
});

const mapDispatchToProps = dispatch => ({
  actions: action => dispatch(action),
});

export default connect(mapStateToprops, mapDispatchToProps)(UpdateArticles);
