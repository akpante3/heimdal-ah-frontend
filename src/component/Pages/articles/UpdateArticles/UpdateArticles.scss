@import '../../../../styles//variables.scss';

.update-articles{
    &.__button {
        float: right;
        font-size: 20px;
        height: 2.5em;
        margin: 2% 7% 0 0;
        width: 11%;
    }

    &.__text {
        font-size: 1.5em;
        margin: 1% 0 15% 0;
        padding: 0 10%;
        text-align: justify;
    }

    &.__title {
        font-size: 3em;
        font-weight: bold;
        margin-top: 11%;
        padding: 0 10%;
        text-align: justify;
    }
}

.medium-editor-element {
    font-family: $font-family__cardo;
    padding-bottom: 2%;
    white-space: normal;
}

.medium-editor-element:focus {
    outline: 0;
}