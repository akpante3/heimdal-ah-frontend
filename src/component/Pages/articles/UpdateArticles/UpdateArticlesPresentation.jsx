import React, { Fragment } from 'react';
import Editor from 'react-medium-editor';
import checkAuthentication from '../../../../utils/checkAuthentication';
import Button from '../../../ui/Buttons/Button';
import Alert from '../../../ui/Alert/Alert';
import './UpdateArticles.scss';
// eslint-disable-next-line import/no-named-as-default
import UpdateArticles from './UpdateArticles';

const UpdatePasswordPresentation = () => (
  <UpdateArticles>
    {
      data => (
        <Fragment>
          {checkAuthentication()}
          {data.updateArticlesStatus === 'SUCCESS' && <Alert type='success' title='Article Update' message='Article updated successfully' />}
          {data.updateArticlesStatus === 'ERROR' && <Alert type='warning' title='Server Error' message='An error occured while trying to update your article, please try again' />}
          <Button type='article' label='Post Article' Class='update-articles __button' disabled onClick={e => data.onHandleSubmit(e)} />
          <div className='update-articles __title'>
            <Editor
              tag="pre"
              data-placeholder="Article Title"
              text={data.title}
              onChange={data.onHandleTitle}
              options={{ toolbar: { buttons: ['bold', 'italic', 'underline', 'anchor', 'h2', 'h3', 'quote', 'html', 'strikethrough', 'subscript', 'superscript'] } }}
            />
          </div>
          {}
          <div className='update-articles __text'>
            <Editor
              tag="pre"
              data-placeholder="Article Body"
              text={data.body}
              onChange={data.onHandleText}
              options={{ toolbar: { buttons: ['bold', 'italic', 'underline', 'anchor', 'h2', 'h3', 'quote', 'html', 'strikethrough', 'subscript', 'superscript'] } }}
            />
          </div>
        </Fragment>
      )
    }
  </UpdateArticles>
);

export default UpdatePasswordPresentation;
