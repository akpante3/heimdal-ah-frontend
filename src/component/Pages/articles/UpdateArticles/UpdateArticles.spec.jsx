/* eslint-disable no-undef */
import React from 'react';
import sinon from 'sinon';
import UpdateArticlesPresentation from './UpdateArticlesPresentation';
import { UpdateArticles } from './UpdateArticles';
import getArticlesById from '../../../../actions/articleActions/getArticlesByIdActions';
import updateArticles from '../../../../actions/articleActions/updateArticlesActions';
import setToken from '../../../../config/setToken';

describe('Update Articles Component Test Suite', () => {
  describe('<UpdateArticlesPresentation />', () => {
    it('should render the UpdateArticlesPresentation component', () => {
      const wrapper = shallow(<UpdateArticlesPresentation />);
      expect(wrapper.length).toBe(1);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the PasswordReset component', () => {
      // eslint-disable-next-line react/no-children-prop
      const wrapper = shallow(<UpdateArticles updateArticlesStatus='SUCCESS' body='body' title='title' children={() => { }} actions={() => 'clicked me'} />);
      expect(wrapper.length).toBe(1);
      expect(wrapper).toMatchSnapshot();
    });

    it('should handle submit events', () => {
      // eslint-disable-next-line react/no-children-prop
      const wrapper = shallow(<UpdateArticles updateArticlesStatus='SUCCESS' body='body' title='title' children={() => { }} actions={() => 'clicked me'} />);
      const updatedTitle = 'updatedTitle';
      const updatedText = 'updatedText';
      const description = 'updatedTitle';
      const identifier = 1;
      const e = { preventDefault: () => { } };

      wrapper.instance().onHandleSubmit(e);
      expect(sinon.stub(jest.fn(wrapper.instance().props.actions(updateArticles(updatedTitle, updatedText, description, identifier)))))
        .not.toHaveBeenCalled();
    });

    it('should handle change events', () => {
      // eslint-disable-next-line react/no-children-prop
      const wrapper = shallow(<UpdateArticles updateArticlesStatus='SUCCESS' body='body' title='title' children={() => { }} actions={() => 'clicked me'} />);
      //   const e = { target: { name: 'email', value: 'Heimdal' }, preventDefault: () => {} };
      expect(wrapper.state().updatedTitle).toEqual('');
      expect(wrapper.state().updatedText).toEqual('');
      wrapper.instance().onHandleText('article text');
      wrapper.instance().onHandleTitle('article title');
      expect(wrapper.instance().state.updatedText).toEqual('article text');
      expect(wrapper.instance().state.updatedTitle).toEqual('article title');
    });

    it('should get article by id', () => {
      // eslint-disable-next-line react/no-children-prop
      const wrapper = shallow(<UpdateArticles updateArticlesStatus='SUCCESS' body='body' title='title' children={() => { }} actions={() => { }} />);
      const identifier = 1;
      localStorage.setItem('access-token', 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlSWQiOiIyIiwidXNlcklkIjoxLCJlbWFpbCI6InVzZXJAaGVpbWRhbC5jb20ifQ.J7HZ-3c4vgOsLS377n5HNeamfdM4xspSgHvIxAHEJE0');
      const token = localStorage.getItem('access-token');
      setToken(token);
      wrapper.instance().componentDidMount();
      expect(jest.fn(getArticlesById(identifier)).mock.calls.length).toBe(0);
    });
  });
});
