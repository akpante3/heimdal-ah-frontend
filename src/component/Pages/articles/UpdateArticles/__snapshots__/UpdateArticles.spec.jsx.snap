// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update Articles Component Test Suite <UpdateArticlesPresentation /> should render the PasswordReset component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UpdateArticles
    actions={[Function]}
    body="body"
    title="title"
    updateArticlesStatus="SUCCESS"
  >
    [Function]
  </UpdateArticles>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [
    undefined,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Update Articles Component Test Suite <UpdateArticlesPresentation /> should render the UpdateArticlesPresentation component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UpdatePasswordPresentation />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
